tag_release.py

```diff
@@ -0,0 +1,554 @@
+    token = None
+    for _ in range(attempts):
+        try:
+            username, token = create_github_creds()
+        except requests.exceptions.RequestException as e:
+            print(
+                "Invalid authentication: {}".format(e.message),
+                file=sys.stderr,
+            )
+            continue
+        else:
+            break
+    if token:
+        print("Successfully authenticated to GitHub", file=sys.stderr)
+    if not token:
+        print("Too many invalid authentication attempts.", file=sys.stderr)
+        return False
+
+    config_file = (path("~/.config") / TOKEN_NAME).expand()
+    # make sure parent directory exists
+    config_file.parent.makedirs_p()
+    # read existing config if it exists
+    if config_file.isfile():
+        with open(config_file) as f:
+            config = json.load(f)
+    else:
+        config = {}
+    # update config
+    if 'credentials' not in config:
+        config["credentials"] = {}
+    if 'api.github.com' not in config['credentials']:
+        config["credentials"]["api.github.com"] = {}
+    config["credentials"]["api.github.com"]["username"] = username
+    config["credentials"]["api.github.com"]["token"] = token
+    # write it back out
+    with open(config_file, "w") as f:
+        json.dump(config, f)
+
+    return True
+
+def openedx_release_repos(session):
+    repos_resp = session.get(REPOS_YAML)
+    repos_resp.raise_for_status()
+    all_repos = yaml.safe_load(repos_resp.text)
+    repos = {name: data for name, data in all_repos.items()
+             if data and data.get("openedx-release")}
+    return repos
+
+
+def commits_to_tag_in_repos(repos, session):
+    to_tag = {}
+    for repo_name, repo_data in repos.items():
+        # make sure the repo exists
+        repo_url = "https://api.github.com/repos/{repo}".format(repo=repo_name)
+        repo_resp = session.get(repo_url)
+        if not repo_resp.ok:
+            print("Invalid repo {repo}".format(repo=repo_name), file=sys.stderr)
+            continue
+        # are we specifying a branch?
+        branch_name = repo_data["openedx-release"].get("branch")
+        if branch_name:
+            try:
+                to_tag[repo_name] = get_latest_commit_for_ref(
+                    repo_name,
+                    branch_name,
+                    session=session,
+                )
+            except RequestException, ValueError:
+                msg = "Invalid branch {branch} in repo {repo}".format(
+                    branch=branch_name,
+                    repo=repo_name,
+                )
+                print(msg, file=sys.stderr)
+            continue
+        # are we specifying a parent repo?
+        parent_repo_name = repo_data["openedx-release"].get("parent-repo")
+        if parent_repo_name:
+            # we need the branch for the parent repo
+            parent_branch_name = repos[parent_repo_name]["openedx-release"]["branch"]
```

Comment:
Need to check that we got a real parent_branch_name.  An error in repos.yaml could have this line raise KeyError.